sql/src/main/java/io/crate/external/S3ClientHelper.java

```diff
@@ -75,39 +68,56 @@ public AWSCredentials getCredentials() {
 
     private final IntObjectMap<AmazonS3> clientMap = new IntObjectHashMap<>(1);
 
-    protected AmazonS3 initClient(@Nullable String accessKey, @Nullable String secretKey) throws IOException {
-        if (accessKey == null || secretKey == null) {
-            return new AmazonS3Client(DEFAULT_CREDENTIALS_PROVIDER_CHAIN, CLIENT_CONFIGURATION);
+
+    public S3ClientHelper(Settings settings) {
+        this.settings = settings;
+    }
+
+    protected AmazonS3 initClient(String accessKey, String secretKey) throws IOException {
+        if (accessKey.isEmpty() || secretKey.isEmpty()) {
+            return AmazonS3ClientBuilder.defaultClient();
         }
-        return new AmazonS3Client(
-            new BasicAWSCredentials(accessKey, secretKey),
-            CLIENT_CONFIGURATION
-        );
+        return AmazonS3ClientBuilder.standard()
+            .withCredentials(new AWSStaticCredentialsProvider(new BasicAWSCredentials(accessKey, secretKey)))
+            .withClientConfiguration(CLIENT_CONFIGURATION)
+            .build();
     }
 
     public AmazonS3 client(URI uri) throws IOException {
-        String accessKey = null;
-        String secretKey = null;
         if (uri.getHost() == null) {
             throw new IllegalArgumentException(INVALID_URI_MSG);
         }
+        // UserInfo should not be part of the URI as the AWS credentials should reside in the keystore
         if (uri.getUserInfo() != null) {
-            String[] userInfoParts = uri.getUserInfo().split(":");
-            try {
-                accessKey = userInfoParts[0];
-                secretKey = userInfoParts[1];
-            } catch (ArrayIndexOutOfBoundsException e) {
-                // ignore
+            if (!uri.getUserInfo().isEmpty()) {
+                throw new IllegalArgumentException("S3 credentials cannot be part of the URI but instead " +
+                                                   "must be stored inside the Crate keystore");
             }
+        } else if (uri.toString().contains("@") && uri.toString().contains(":")) {
             // if the URI contains '@' and ':', a UserInfo is in fact given, but could not
             // be parsed properly because the URI is not valid (e.g. not properly encoded).
-        } else if (uri.toString().contains("@") && uri.toString().contains(":")) {
             throw new IllegalArgumentException(INVALID_URI_MSG);
         }
-        return client(accessKey, secretKey);
+
+        return client();
     }
 
-    private AmazonS3 client(@Nullable String accessKey, @Nullable String secretKey) throws IOException {
+    private AmazonS3 client() throws IOException {
+        String accessKey = ACCESS_KEY_SETTING.getConcreteSetting(ACCESS_KEY_SETTING_NAME).get(settings).toString();
+        String secretKey = SECRET_KEY_SETTING.getConcreteSetting(SECRET_KEY_SETTING_NAME).get(settings).toString();
+
+        if (accessKey.isEmpty() || secretKey.isEmpty()) {
+            if (!accessKey.isEmpty()) {
```

Comment:
I'd just throw one exception if any is empty. I don't think it adds much for the user to differ here.

Reply:
ok!